tests/test_ci_contract.py

```
@@ -14,12 +14,78 @@
+    for raw in path.read_text(encoding="utf-8").splitlines():
+        line = raw.strip()
+        if line and not line.startswith("#"):
+            parsed.append(Requirement(line))
+    return parsed
+
+
+def _exact_pin(requirement: Requirement) -> str:
+    """Return an exact pinned version or reject the requirement."""
+    specs = list(requirement.specifier)
+    if len(specs) != 1 or specs[0].operator != "==":
+        raise AssertionError(f"CI dependency must be exactly pinned: {requirement}")
```

Ruff (0.16.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)